homeassistant/components/event/__init__.py

```diff
@@ -0,0 +1,209 @@
+from homeassistant.helpers.config_validation import (  # noqa: F401
+    PLATFORM_SCHEMA,
+    PLATFORM_SCHEMA_BASE,
+)
+from homeassistant.helpers.entity import EntityDescription
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.restore_state import ExtraStoredData, RestoreEntity
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.util import dt as dt_util
+
+from .const import ATTR_EVENT_TYPE, ATTR_EVENT_TYPES, DOMAIN
+
+SCAN_INTERVAL = timedelta(seconds=30)
+
+ENTITY_ID_FORMAT = DOMAIN + ".{}"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class EventDeviceClass(StrEnum):
+    """Device class for events."""
+
+    DOORBELL = "doorbell"
+    BUTTON = "button"
+    MOTION = "motion"
+
+
+__all__ = [
+    "ATTR_EVENT_TYPE",
+    "ATTR_EVENT_TYPES",
+    "DOMAIN",
+    "PLATFORM_SCHEMA_BASE",
+    "PLATFORM_SCHEMA",
+    "EventDeviceClass",
+    "EventEntity",
+    "EventEntityDescription",
+    "EventEntityFeature",
+]
+
+# mypy: disallow-any-generics
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up Event entities."""
+    component = hass.data[DOMAIN] = EntityComponent[EventEntity](
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+    await component.async_setup(config)
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    component: EntityComponent[EventEntity] = hass.data[DOMAIN]
+    return await component.async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    component: EntityComponent[EventEntity] = hass.data[DOMAIN]
+    return await component.async_unload_entry(entry)
+
+
+@dataclass
+class EventEntityDescription(EntityDescription):
+    """A class that describes event entities."""
+
+    device_class: EventDeviceClass | None = None
+    event_types: list[str] | None = None
+
+
+@dataclass
+class EventExtraStoredData(ExtraStoredData):
+    """Object to hold extra stored data."""
+
+    last_event_type: str | None
+    last_event_attributes: dict[str, Any] | None
+
+    def as_dict(self) -> dict[str, Any]:
+        """Return a dict representation of the event data."""
+        return asdict(self)
+
+    @classmethod
+    def from_dict(cls, restored: dict[str, Any]) -> Self | None:
+        """Initialize a stored event state from a dict."""
+        try:
+            return cls(
+                restored["last_event_type"],
+                restored["last_event_attributes"],
+            )
+        except KeyError:
+            return None
+
+
+class EventEntity(RestoreEntity):
+    """Representation of a Event entity."""
+
+    entity_description: EventEntityDescription
+    _attr_device_class: EventDeviceClass | None
+    _attr_event_types: list[str]
+    _attr_state: None
+
+    __last_event: datetime | None = None
```

Comment:
Let's be clearer since there are other `__last_event_xxx`:
```suggestion
    __last_event_timestamp: datetime | None = None
```
Or:
```suggestion
    __last_event_fired: datetime | None = None
```

Reply:
Done in [caf3e58](https://github.com/home-assistant/core/pull/96797/commits/caf3e5828a807fb581a55d642df48956a7ae85ca)